### DISCUSSION TITLE
[DEV] modules data table

### DISCUSSION THREAD

**Discussion by FrancescoBorzi** [azerothcore/azerothcore-wotlk#3916] [General]
It was discussed some time ago the possibility of including in AC a table to let modules store persistent data. No action has been taken so far, so I open this thread in order to bring back the discussion.

Comment from @Deku :

> I think that just adding a table with the following should be a good start:
> - module name or id
> - character guid
> - data (as json/XML/yaml)
> - creation_date (just for auditing purposes)
> that way, stuff like custom rates or other per-character custom configs can me stored and used across play sessions
> (and you don't fill your database with many tables)
> 

Comment from me:

> I like the idea!
> About the table structure, using json or similar format for the data would make it more compact, on the other hand would make it harder to look and filter  for specific data (e.g. For statistic tools)
> Another option could be:
> - character guid
> - data_name (e. g. 'shinmodule_data1')
> - value
> - creation_date (or timestamp for last edit date)
> And the unique key would be the first two columns
> so if my module needs 2 store two (unrelated) values per each character, whatever they mean, every character in the table will have 2 entries. One having data_name = shinmodule_data1 and data_name = shinmodule_data2 with their values
> it's less compact but more handy to query
> or we could make it more explicit to avoid conflicts in different modules and split data_name into two tables: module and key
> so it would be character_guid, module, key, value, timestamp
> UNIQUE KEY is the first three in this case

What are the opinions/suggestions about this?

**Comment by brussens** [comment]
The module is first of all a logically complete independent entity. Therefore, it makes no sense to bind each module to a certain structure in the database.

**Comment by BarbzYHOOL** [comment]
storing the data in yaml, json etc in the DB is a nono for me
